---
title: "cim22d-atlas seurat4"
output: html_document
author: "chenyu"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  cache = TRUE,
  cache.lazy = FALSE,
  tidy = TRUE)
```



### Setup the Seurat Object

```{r init, message=FALSE}
library(Seurat)
library(dplyr)

# Load the root single cells dataset
# 将counts, genes/Fearture, barcode等信息综合为矩阵，row为基因，column为细胞
cim22d.raw <- Read10X(data.dir = "./rawdata/filtered_feature_bc_matrix/")

# Examine the memory savings between regular and sparse matrices
dense.size <- object.size(x = as.matrix(x = cim22d.raw))
dense.size

sparse.size <- object.size(x = cim22d.raw)
sparse.size

dense.size / sparse.size
```


```{r create.object, results='hide', message=FALSE}
cim22d <- CreateSeuratObject(counts = cim22d.raw, 
                             min.cells = 3, 
                             min.features = 200, 
                             project = "10X_cim22d")
```


### QC and selecting cells for further analysis


```{r qc, results='hide', fig.height=7,fig.width=13}
# 获取线粒体基因,mito.features为线粒体基因名
mito.features <- grep(pattern = "^LOC-Osm", 
                      x = rownames(x = cim22d), 
                      value = TRUE)

# 计算每个cell中，线粒体基因counts数目/总基因counts数
percent.mito <- Matrix::colSums(x = GetAssayData(object = cim22d, slot = 'counts')[mito.features, ]) / Matrix::colSums(x = GetAssayData(object = cim22d, slot = 'counts'))

# 添加线粒体比例，方便后续过滤
cim22d[['percent.mito']] <- percent.mito

# 画图展示
pdf(file = "./plot/01_nFeature_nCount.pdf",width = 10,height = 10)
plot(VlnPlot(object = cim22d, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mito"), 
        ncol = 3))
dev.off()
```



```{r, results='hide', fig.height=7,fig.width=13}
FeatureScatter(object = cim22d, feature1 = "nCount_RNA", feature2 = "percent.mito")
```



```{r,fig.height=7,fig.width=13}
FeatureScatter(object = cim22d, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
```


```{r,results='hide', fig.height=7,fig.width=13 }
# We filter out cells that have unique feature counts over 2,500 or less than 200
cim22d_hvg1 <- subset(x = cim22d, 
                 subset = nFeature_RNA > 500 & nFeature_RNA < 5000 & percent.mito < 0.05)

# 过滤后可以展示过后基因和表达量
FeatureScatter(object = cim22d_hvg1,feature1 = "nCount_RNA",feature2 = "nFeature_RNA")
```



### Normalizing the data

```{r normalize, results='hide'}
cim22d_hvg1 <- NormalizeData(object = cim22d_hvg1, 
                        normalization.method = "LogNormalize", 
                        scale.factor = 1e4)
```



### Detection of variable features across the single cells


```{r var_genes}
cim22d_hvg1 <- FindVariableFeatures(object = cim22d_hvg1, 
                               selection.method = 'mean.var.plot',
                               mean.cutoff = c(0.0125, 3), 
                               dispersion.cutoff = c(1.0, Inf))


#画图
VariableFeaturePlot(object = cim22d_hvg1)

length(x = VariableFeatures(object = cim22d_hvg1))
```


考虑转录因子和细胞周期基因影响

```{r }
at.CC.gene <- read.csv("./rawdata/core_cellcycle_gene_os.csv",header = TRUE)
at.TF.gene <- read.csv("rawdata/OsTF_RAP_unique.csv",header = TRUE)
at.CC.gene$rice.gene.locus.identifier %in% VariableFeatures(cim22d_hvg1)
sum(at.TF.gene$OsTF_RAP_unique %in% VariableFeatures(cim22d_hvg1))
```



### Scaling the data and removing unwanted sources of variation


```{r regress, fig.height=7, fig.width=11, results='hide'}
cim22d_hvg1 <- ScaleData(object = cim22d_hvg1, 
                    features = rownames(x = cim22d_hvg1), 
                    vars.to.regress = c("nCount_RNA", "percent.mito"))

```




### Perform linear dimensional reduction

```{r pca}
cim22d_hvg1 <- RunPCA(object = cim22d_hvg1, 
                 features = VariableFeatures(object = cim22d_hvg1), 
                 verbose = FALSE,
                 npcs = 100)
```


使用VizDimReduction, DimPlot, and DimHeatmap展示PCA结果
显示前5个主成分和每个主成分前5个基因。
```{r pca_viz}
print(x = cim22d_hvg1[['pca']], 
      dims = 1:5, 
      nfeatures = 5, 
      projected = FALSE)
```


VizDimReduction展示线性降纬数据：
```{r }
pdf("./plot/cim22d_hvg1_LDR1.pdf",width = 10,height = 10)
plot(VizDimLoadings(object = cim22d_hvg1, dims = 1:2))
dev.off()
```

DimPlot展示降纬结果：
```{r }
pdf("./plot/cim22d_hvg2_LDR2.pdf",width = 10,height = 10)
DimPlot(object = cim22d_hvg1)
dev.off()
```

DimHeatmap展示降纬结果：
```{r }
cim22d_hvg1 <- ProjectDim(object = cim22d_hvg1)

#展示纬度1的数据
DimHeatmap(object = cim22d_hvg1, dims = 1, cells = 500, balanced = TRUE)
#展示纬度1～20的数据
DimHeatmap(object = cim22d_hvg1, dims = 1:20, cells = 500, balanced = TRUE)
```





### Determine statistically significant principal components


```{r jackstraw, fig.height=6, fig.width=10, warning=FALSE}
cim22d_hvg1 <- JackStraw(object = cim22d_hvg1, num.replicate = 100)
cim22d_hvg1 <- ScoreJackStraw(object = cim22d_hvg1, dims = 1:20)
```

展示分析结果
```{r jsplots, fig.height=6, fig.width=10, warning=FALSE}
JackStrawPlot(object = cim22d_hvg1, dims = 1:20)
```

进一步判断有效主成分
```{r elbow_plot, fig.height=6, fig.width=10, warning=FALSE}
ElbowPlot(object = cim22d_hvg1)
```



### Cluster the cells

```{r dimension}
dimension <- c(1:40)
res = 1.0
```


```{r cluster, fig.height=5, fig.width=7}
cim22d_hvg1 <- FindNeighbors(object = cim22d_hvg1, dims = dimension)
cim22d_hvg1 <- FindClusters(object = cim22d_hvg1, resolution = res)
```




### Run Non-linear dimensional reduction (tSNE)

```{r tsne, fig.height=5, fig.width=7, message=FALSE, warnings=FALSE}
cim22d_hvg1 <- RunTSNE(object = cim22d_hvg1, dims = 1:20)
```


```{r tsneplot, fig.height=5, fig.width=7}
pdf(file = "./plot/cim22d_hvg1_d40_tSNE.pdf",width = 10,height = 10)
plot(DimPlot(object = cim22d_hvg1, reduction = 'tsne',label = TRUE,pt.size = 1))
dev.off()
```



### Run UMAP

添加Python路径

```{r }
# load Python
library(reticulate)


cim22d_hvg1 <- RunUMAP(object = cim22d_hvg1, 
                         reduction = "pca", 
                        umap.method = "uwot",
                         dims = 1:20, 
                         n.neighbors = 30L,
                         metric = "correlation",
                         min.dist = 0.3)

pdf(file = "./plot/cim22d_d40_umap.pdf",width = 10,height = 10)
plot(DimPlot(object = cim22d_hvg1, reduction = "umap", label = TRUE, repel = TRUE,pt.size = 1))
dev.off()

```



### 使用已知gene评价聚类效果
RAP              | Symbol
--------------   |----------
Os01g0660200     | NA vascular cell
Os01g0134900     | NA vascular bundle
Os05g0349800     | OsEm1 proliferating cell
Os03g0699000     | OLE18 proliferating cell
Os06g0160100     | NA cell cycle cell
Os01g0971000     | NA vascular bundle
Os06g0653000     | OsGL1-3 epidermis
Os03g0184500     | B3 initial cell
Os01g0896400     | NA initial cell
Os06g0127800     | DLT initial cell
Os04g0105200     | NA seive element
Os03g0135100     | NA seive element




```{r known.marker}
known.markers <- c("Os01g0660200","Os01g0134900","Os05g0349800","Os03g0699000","Os06g0160100","Os01g0971000","Os06g0653000","Os03g0184500","Os01g0896400","Os06g0127800","Os04g0105200","Os03g0135100")
```

```{r markerplots, fig.height=8, fig.width=15,}
pdf("./plot/cim22d_hvg1_d40_known_makers.pdf",width = 20,height = 20)
plot(VlnPlot(object = cim22d_hvg1, features = known.markers))
dev.off()
```


### Save the object.
```{r saveobject}
saveRDS(cim22d_hvg1, file = "cim22d_hvg1.rds")
```




### Finding differentially expressed genes (cluster biomarkers)


```{r markers1, fig.height=8, fig.width=15, warning=FALSE}
# find all markers of cluster 1
cluster1.markers_cim22d_hvg1 <- FindMarkers(object = cim22d_hvg1, 
                                              ident.1 = 1, 
                                              min.pct = 0.25)
head(x = cluster1.markers_cim22d_hvg1, n = 5)
```


查找cluster 0 maker genes
```{r }
cluster0.markers_cim22d_hvg1 <- FindMarkers(object = cim22d_hvg1, 
                                              ident.1 = 0, 
                                              min.pct = 0.25)
head(x = cluster0.markers_cim22d_hvg1, n = 5)
```



```{r }
# find all markers distinguishing cluster 5 from clusters 0 and 3
cluster5.markers_cim22d_hvg1 <- FindMarkers(object = cim22d_hvg1, 
                                ident.1 = 5, 
                                ident.2 = c(0, 3), 
                                min.pct = 0.25)
head(x = cluster5.markers_cim22d_hvg1, n = 5)
```

查找所有显著cell marker genes(seraut4把avg_logFC变成avg_log2FC)
```{r }
#find markers for every cluster compared to all remaining cells, report only the positive ones
cim22d_hvg1.markers <- FindAllMarkers(object = cim22d_hvg1, 
                                        only.pos = TRUE, 
                                        min.pct = 0.25, 
                                        logfc.threshold = 0.25)
cim22d_hvg1.markers %>% group_by(cluster) %>% top_n(n = 2, wt = avg_log2FC)

```


各Cluster中Top10基因表达heatmap(seraut4中avg_logFC变成了avg_log2FC)
```{r }
cim22d_hvg1.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC) -> top10

pdf(file = "./plot/cim22d_hvg1_top10.pdf",width = 10,height = 10)
plot(DoHeatmap(object = cim22d_hvg1, features = top10$gene) + NoLegend())
dev.off()
```




### 添加注释（该版本的注释不是很多，20220616没有用这个）
```{r }
#install.packages("/data3/wanglab/yc/reference/org.Osativa.eg.db.tar.gz", repo=NULL)
library(org.Osativa.eg.db)

cim22d.markers$Symbol <- mapIds(org.Osativa.eg.db, keys = cim22d.markers$gene, 
                                  column = "RAP_Symbol",
                                  multiVals = "first",
                                  keytype = "GID", )

cim22d.markers$Name <- mapIds(org.Os2.eg.db, keys = cim22d.markers$gene, 
                                  column = "RAP_Name",
                                  multiVals = "first",
                                  keytype = "GID", )




write.csv(x=cim22d.markers,file = "./result/cim22d_marker_genes.csv")
```

###anotation from wanmuchun
```{r}
anno <- read.csv("~/reference/Oryza_sativa/annotation/IRGSP_OS2tair_anno.csv")

cim22d_hvg1.markers <- left_join(cim22d_hvg1.markers, anno, by=c("gene"="RAP_ID"))

write.csv(x=cim22d_hvg1.markers,file = "./result/cim22d_marker_genes_bywmcano.csv")


```

```{r }
cluster1.markers_cim22d_hvg1 <- FindMarkers(object = cim22d_hvg1, 
                                ident.1 = 0, 
                                logfc.threshold = 0.25, 
                                test.use = "roc", 
                                only.pos = TRUE)
```



### 自定义基因Marker Gene画图展示

```{r }
input.genes <- c("Os01g0660200","Os01g0134900","Os05g0349800","Os03g0699000","Os06g0160100","Os01g0971000","Os06g0653000","Os03g0184500","Os01g0896400","Os06g0127800","Os04g0105200","Os03g0135100")
```

```{r markerplots, fig.height=8, fig.width=15,}
VlnPlot(object = cim22d_hvg1, 
        features = input.genes)
```


```{r }
# you can plot raw counts as well
VlnPlot(object = cim22d_hvg1, 
        features = input.genes,
        slot = 'counts', 
        log = TRUE)
```


```{r }
FeaturePlot(object = cim22d_hvg1, 
            features = input.genes)
```


### 添加注释（TQ师兄的）
```{r }
install.packages("/data3/wanglab/yc/reference/org.Os2.eg.db", repo=NULL)
library(org.Os2.eg.db)

cim22d_hvg1.markers$Symbol <- mapIds(org.Os2.eg.db, keys = cim22d_hvg1.markers$gene, 
                                  column = "RAP_Symbol",
                                  multiVals = "first",
                                  keytype = "GID", )

cim22d_hvg1.markers$Name <- mapIds(org.Os2.eg.db, keys = cim22d_hvg1.markers$gene, 
                                  column = "RAP_Name",
                                  multiVals = "first",
                                  keytype = "GID", )




write.csv(x=cim22d_hvg1.markers,file = "./result/cim22d_marker_genes_byTQanno.csv")
```


```{r}
tmp <- read.csv("./result/cim22d_marker_genes_byTQanno.csv", row.names = 1)

out <- lapply(unique(tmp$cluster), function(x) {
    tmp[tmp$cluster == x, ]
})

names(out) <- paste0("cluster_", unique(tmp$cluster))

openxlsx::write.xlsx(out, "./result/cim22d_markers_bywmc_addTQ.xlsx")
```























